src/linear_algebra/affine_space.lean

```diff
@@ -0,0 +1,206 @@
+{p | ∃ p1 ∈ s, ∃ v ∈ (vector_span k V s).carrier, p = v +ᵥ p1}
+
+/-- The set of points in the affine span of a nonempty set of points
+is nonempty. -/
+lemma span_points_nonempty_of_nonempty {s : set P} (h : s.nonempty) :
+  (span_points k V s).nonempty :=
+begin
+  cases h with p hp,
+  use p,
+  unfold span_points,
+  use p,
+  use hp,
+  use 0,
+  exact and.intro (submodule.zero _) (zero_vadd V p).symm
+end
+
+/-- Adding a point in the affine span and a vector in the spanning
+subspace produces a point in the affine span. -/
+lemma vadd_mem_span_points_of_mem_span_points_of_mem_vector_span {s : set P} {p : P} {v : V}
+    (hp : p ∈ span_points k V s) (hv : v ∈ vector_span k V s) : v +ᵥ p ∈ span_points k V s :=
+begin
+  unfold span_points at hp,
+  cases hp with p2 hp2,
+  cases hp2 with hp2 hp3,
+  rcases hp3 with ⟨v2, ⟨hv2, hv2p⟩⟩,
+  rw [hv2p, vadd_assoc],
+  unfold span_points,
+  use p2,
+  use hp2,
+  use v + v2,
+  exact and.intro ((vector_span k V s).add hv hv2) rfl
+end
+
+/-- Subtracting two points in the affine span produces a vector in the
+spanning subspace. -/
+lemma vsub_mem_vector_span_of_mem_span_points_of_mem_span_points {s : set P} {p1 p2 : P}
+    (hp1 : p1 ∈ span_points k V s) (hp2 : p2 ∈ span_points k V s) :
+  p1 -ᵥ p2 ∈ vector_span k V s :=
+begin
+  unfold span_points at hp1 hp2,
+  cases hp1 with p1a hp1,
+  cases hp2 with p2a hp2,
+  cases hp1 with hp1a hp1b,
+  cases hp2 with hp2a hp2b,
+  rcases hp1b with ⟨v1, ⟨hv1, hv1p⟩⟩,
+  rcases hp2b with ⟨v2, ⟨hv2, hv2p⟩⟩,
+  rw [hv1p, hv2p, vsub_vadd_eq_vsub_sub V (v1 +ᵥ p1a), vadd_vsub_assoc, add_comm,
+      add_sub_assoc],
+  have hv1v2 : v1 - v2 ∈ (vector_span k V s).carrier,
+  { apply (vector_span k V s).add hv1,
+    rw ←neg_one_smul k v2,
+    exact (vector_span k V s).smul (-1 : k) hv2 },
+  refine (vector_span k V s).add _ hv1v2,
+  unfold vector_span,
+  change p1a -ᵥ p2a ∈ submodule.span k (vsub_set V s),
+  have hp1p2 : p1a -ᵥ p2a ∈ vsub_set V s,
+  { unfold vsub_set,
+    use p1a,
+    use hp1a,
+    use p2a,
+    use hp2a },
+  have hp1p2s : vsub_set V s ⊆ submodule.span k (vsub_set V s) := submodule.subset_span,
+  apply set.mem_of_mem_of_subset hp1p2 hp1p2s
+end
+
+end affine_space
+
+open add_torsor affine_space
+
+section affine_subspace
+
+variables (k : Type*) (V : Type*) (P : Type*) [field k] [add_comm_group V] [vector_space k V]
+          [S : affine_space k V P]
+include S
+
+/-- An `affine_subspace k V P` is a subset of an `affine_space k V P`
+that has an affine space structure induced by a corresponding subspace
+of the `vector_space k V`. -/
+structure affine_subspace :=
+(carrier : set P)
+(direction : subspace k V)
+(nonempty : carrier.nonempty)
+(add : ∀ (p : P) (v : V), p ∈ carrier → v ∈ direction.carrier → v +ᵥ p ∈ carrier)
+(sub : ∀ (p1 p2 : P), p1 ∈ carrier → p2 ∈ carrier → p1 -ᵥ p2 ∈ direction.carrier)
+
+/-- The whole affine space as a subspace of itself. -/
+def affine_subspace.of_affine_space : affine_subspace k V P :=
```

Comment:
Is this name following a pattern from somewhere else? I'm not entirely up on the naming conventions in this corner of the library, but this sounds a little funny. We're already in the `affine_space` namespace, and projection notation won't be useful. `univ_affine_subspace` or something like that?

Reply:
Renamed as suggested.